text/0000-transparent-unions.md

```diff
@@ -0,0 +1,103 @@
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+It would be nice to make `MaybeUninit<T>` `#[repr(transparent)]`. This type is a `union`, and thus this RFC is required in order to allow making it transparent.
+
+Of course, the standard "do nothing" alternative exists. Rust doesn't strictly *require* this feature. But it would benefit from this, so the "do nothing" alternative is undesirable.
+
+# Prior art
+[prior-art]: #prior-art
+
+See [the discussion on RFC #1758](https://github.com/rust-lang/rfcs/pull/1758) (which introduced `#[repr(transparent)]`) for some discussion on applying the attribute to a `union`. A summary of the discussion:
+
+> https://github.com/rust-lang/rfcs/pull/1758#discussion_r80436621
+> **nagisa:** "Why not univariant unions and enums?"
+> **nox:** "I tried to be conservative for now given I don't have a use case for univariant unions and enums in FFI context."
+
+> https://github.com/rust-lang/rfcs/pull/1758#issuecomment-254872520
+> **eddyb:** "I found another important usecase: for `ManuallyDrop<T>`, to be useful in arrays (i.e. small vector optimizations), it needs to have the same layout as `T` and AFAICT `#[repr(C)]` is not guaranteed to do the right thing"
+> **retep998:** "So we'd need to be able to specify `#[repr(transparent)]` on unions?"
+> **eddyb:** "That's the only way to be sure AFAICT, yes."
+
+> https://github.com/rust-lang/rfcs/pull/1758#issuecomment-274670231
+> **joshtriplett:** "In terms of interactions with other features, I think this needs to specify what happens if you apply it to a union with one field, a union with multiple fields, a struct (tuple or otherwise) with multiple fields, a single-variant enum with one field, an enum struct variant where the enum uses `repr(u32)` or similar. The answer to some of those might be "compile error", but some of them (e.g. the union case) may potentially make sense in some contexts."
+
+> https://github.com/rust-lang/rfcs/pull/1758#issuecomment-290757356
+> **pnkfelix:** "However, I personally do not think we need to expand the scope of the feature. So I am okay with leaving it solely defined on `struct`, and leave `union`/`enum` to a follow-on RFC later. (Much the same with a hypothetical `newtype` feature.)"
+
+In summary, many of the questions regarding `#[repr(transparent)]` on a `union` were the same as applying it to a multi-field `struct`. These questions have since been answered, and I see no problems with applying those same answers to `union`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+None (yet).
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Univariant `enum`s are ommitted from this RFC in an effort to keep the scope small and avoid unnecessary bikeshedding. A future RFC could explore `#[repr(transparent)]` on a univariant `enum`.
```

Comment:
The RFC states in the motivation that:
> ...this RFC rounds out the `#[repr(transparent)]` feature.

so you are making an argument by consistency, but that story is not completed unless we also support `enum`s.

Moreover, the extension of `#[repr(transparent)]` to `enum`s should have completely predictable static and dynamic semantics: Univariant enums are supported and treated as structs. Thus, I don't understand why it would lead to "bikeshedding" at all.

Reply:
I've revised the text to say "a step towards rounding out the `#[repr(transparent)]` feature" and have removed the inflammatory "and avoid unnecessary bikeshedding" text.

Reply:
To clarify: I think this RFC should just deal with univariant enums now instead of postponing them for the future. Conservatism is not an end in itself and I think we should make the language consistent sooner rather than revisiting later so that we can be done with `#[repr(transparent)]`.

Reply:
The only question I have about transparent univariant enums is whether ZSTs are permitted to be transparent or not. Currently structs cannot ([even if they only have a single field](https://rust.godbolt.org/z/KMZpGF)).

Reply:
After thinking on this for a day, I've concluded that no, ZST univariant enums should not be permitted to be transparent (unless they're generic over `T` and have a field of type `T`). This follows the same rules for structs, and makes sense since a univariant enum is effectively just a struct. They shouldn't be treated specially or differently.

I'll revise the RFC to include univariant enums.

Reply:
Okay, I have now updated the RFC to include univariant enums.

Reply:
Much obliged!